pkg/build/util/util.go

```diff
@@ -307,3 +307,41 @@ func SafeForLoggingS2IConfig(config *s2iapi.Config) *s2iapi.Config {
 	newConfig.ScriptsURL, _ = s2iutil.SafeForLoggingURL(newConfig.ScriptsURL)
 	return &newConfig
 }
+
+// GetBuildEnv gets the build strategy environment
+func GetBuildEnv(build *buildapi.Build) *[]kapi.EnvVar {
+	switch {
+	case build.Spec.Strategy.SourceStrategy != nil:
+		return &build.Spec.Strategy.SourceStrategy.Env
+	case build.Spec.Strategy.DockerStrategy != nil:
+		return &build.Spec.Strategy.DockerStrategy.Env
+	case build.Spec.Strategy.CustomStrategy != nil:
+		return &build.Spec.Strategy.CustomStrategy.Env
+	case build.Spec.Strategy.JenkinsPipelineStrategy != nil:
+		return &build.Spec.Strategy.JenkinsPipelineStrategy.Env
+	default:
+		return nil
+	}
+}
+
+// UpdateBuildEnv updates the strategy environment
+// This will replace the existing variable definitions with provided env
+func UpdateBuildEnv(build *buildapi.Build, env []kapi.EnvVar) {
```

Comment:
```
existing strategy envs: A=1, B=2, C=3, D=$(B)
buildrequest envs: B=$(C)
resulting envs: A=1, B=$(C), C=3, D=$(B)
evaluates after substitution to: A=1, B=$(C), C=3, D=$(C)
```

strictly speaking, the resulting envs (before substitution) will be:

```
resulting envs: A=1, C=3, D=$(B), B=$(C)
```
(the env vars from the buildrequest are effectively appended to the existing build envs)

but regardless you'll have an ordering problem.  to be fair, it's an ordering problem that you can also create just by defining your vars in the wrong order in a buildconfig, though.

I think the net is when you're using env var subtitution like this, "here be dragons" and yeah, there will be cases where people want to do substitution operations that just aren't possible because we have to pick an order for processing them (either putting the buildrequest envs first, or last).